src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,254 @@
+    kernel_start_print(buffer);
+    PrintFuture {}
+}
+
+// New! The callback the kernel calls.
+extern fn callback() {
+    // TODO: Wake up the currently-waiting PrintFuture.
+}
+```
+
+So far so good. Barbara then works on `poll`:
+
+```rust
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        if kernel_is_print_done() {
+            Poll::Ready(kernel_get_buffer_back())
+        }
+        Poll::Pending
+    }
+```
+
+Of course, there's something missing here. How does the callback wake the
+`PrintFuture`? She needs to store the
+[`Waker`](https://doc.rust-lang.org/stable/core/task/struct.Waker.html)
+somewhere! Barbara puts the `Waker` in a global variable so the callback can
+find it (this is fine because the app is single threaded and callbacks do NOT
+interrupt execution the way Unix signals do):
+
+```rust
+static mut PRINT_WAKER: Option<Waker> = None;
+
+extern fn callback() {
+    if let Some(waker) = unsafe { PRINT_WAKER.as_ref() } {
+        waker.wake_by_ref();
+    }
+}
+```
+
+She then modifies `poll` to set `PRINT_WAKER`:
+
+```rust
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        if kernel_is_print_done() {
+            return Poll::Ready(kernel_get_buffer_back());
+        }
+        unsafe { PRINT_WAKER = Some(cx.waker()); }
+        Poll::Pending
+    }
+```
+
+`PRINT_WAKER` is stored in `.bss`, which occupies space in RAM but not flash. It
+is two words in size. It points to a
+[`RawWakerVTable`](https://doc.rust-lang.org/stable/core/task/struct.RawWakerVTable.html)
+that is provided by the executor. `RawWakerVTable` does not appear to be
+designed for a no-`alloc` environment. In a no-`alloc` environment, `drop` and
+`clone` are generally no-ops, and `wake`/`wake_by_ref` seem like duplicates.
+Looking at `RawWakerVTable` made Barbara think about the size impact of using
+`Future`.
+
+Barbara decided the overhead would be worth it if it made it easy to work with
+future combinators from external crates *and* could function as a lightweight
+form of multithreading. To support this, Barbara decides to implement an
+executor designed to function like a background thread. Because `alloc` is not
+available, this executor contains a single future. The executor has a `spawn`
+function that accepts a future and starts running that future (overwriting the
+existing future in the executor if one is already present). Once started, the
+executor runs entirely in kernel callbacks.
+
+Barbara identifies several factors that add branching and error handling code to
+the executor:
+
+1. `spawn` should be a safe function, because it is called by high-level
+   application code. However, that means it can be called by the future it
+   contains. If handled naively, this would result in dropping the future while
+   it executes. Barbara adds runtime checks to identify this situation.
+1. `Waker` is `Sync`, so on a multithreaded system, a future could give another
```

Comment:
As outlined earlier, this is some kind of semi-legit design. The `Waker`s really should just schedule the application tasks again and not trying to poll inline.

Reply:
I'm having trouble understanding you here. Are you suggesting something like the following?

1. `wake` marks a `Future` as ready to run, but does not call `poll`.
2. The executor offers a `poll_future` function that polls the future if the future has been awoken (repeatedly, if necessary).
3. The application code has to call `poll_future` after invoking kernel callbacks.

Then the application would need to have a main loop, and that loop would look something like the following:

```rust
loop {
    TASK1.poll_future();
    TASK2.poll_future();
    TASK3.poll_future();
    kernel_ready_for_callbacks();
}
```

Reply:
Yes. But the main-loop would typically be something like `executor.run()` - and you've spawned all your tasks on the executor upfront.

Reply:
After all something like a Winapi/QT/GTK eventloop are also executors :-)

Reply:
I see how that would work, but the benefit is not clear to me.

Reply:
I suppose that you could make all `RawWaker` methods no-ops and instead poll every unfinished `Future` on every iteration, saving some code in the executor at the cost of runtime performance.

Reply:
This is e.g. required to avoid the reentrancy problems. Let's say our async task calls a UART driver, which submits some bytes. It only writes half of it as part of an `async fn write(&self, &[u8]) -> Result<usize, Error>`. The amounts of bytes written is signalled by the interrupt context, and then forwarded as the result of this sub-`Future` - which returns `Ready`. From there another call to the UART driver is made to start another async operation. This operation is now started from an interrupt context instead of the normal application context. In usual environments that is not allowed, because it now requires that resource to be reentrancy safe. 

When the usual `Future` model is implemented it isn't required to be reentrancy safe, because the execution model already guarantees that from inside a callback no further operations are started. The only thing which will ever happen is waking up the task in the executor, so that it can later run from a safe context. 

This also prevents some other issues, e.g. "what happens if this resource is freed inside the callback". But maybe that's not too important in your "everything is static and will never get freed" scenario.

Reply:
> I suppose that you could make all RawWaker methods no-ops and instead poll every unfinished Future on every iteration, saving some code in the executor at the cost of runtime performance.

Yes, that is indeed the most basic executor. 

More efficient versions can e.g. block the main thread using an OS synchronization mechanism or a low level one like `WFI`, and wake the executor from the interrupt.

https://docs.rs/embedded-executor/0.6.1/embedded_executor/ seems like a decent example on what can be done.

Reply:
Regarding the dynamic allocs there: You would avoid them by making sure they only run at startup, and size the memory pool which is used for it accordingly (yes, requires some fiddling :'( ). For the list of ready tasks, it's probably doable with a fixed size list or even some bitmasks if a well known amount of tasks need to be executed.